notes/error.md

```diff
@@ -1,349 +1,444 @@
+if and only if there exists a value $$M$$ and some $$h_0$$ such that
+$$|f(h)| \leq M|g(h)|$$ $$\forall h$$ where $$0 < h < h_0$$
+
+
+But what if we want to consider the function approaching an arbitrary value?
+Then we can redefine the expression as:
+
+
+Let $$f$$ and $$g$$ be two functions. Then
+$$f(x) = \mathcal{O}(g(x))$$ as $$x \rightarrow a$$
+if and only if there exists a value $$M$$ and some $$\delta$$ such that
+$$|f(x)| \leq M|g(x)|$$ $$\forall x$$ where $$0 < |x − a| < \delta$$
+
+For example, consider the function $$f(x) = 2x^2 + 27x + 1000$$
+
+When $$x \rightarrow \infty$$, the term $$x^2$$ is the most significant, and hence $$f(x) = O(x^2)$$
+
+## Big-O Examples - Time Complexity
+
+We can use Big-O to describe the time complexity of our algorithms.
+
+Consider the case of matrix-matrix multiplication.
+If the size of each of our matrices is $$n \times n$$,
+then the time it will take to multiply the matrices is $$\mathcal{O}(n^3)$$ meaning
+that $$\text{Run time} \approx C \cdot n^3$$.  Suppose we know that for $$n_1=1000$$,
+the matrix-matrix multiplication takes 5 seconds.  Estimate how much time
+it would take if we double the size of our matrices to $$2n \times 2n$$.
+
+We know that:
+
+
+$$\begin{align*}
+\text{Time}(2n_1) &\approx C \cdot (2n_1)^3 \\
+&= C \cdot 2^3 \cdot n_1^3\\
+&= 8 \cdot (C \cdot n_1^3) \\
+&= 8 \cdot \text{Time}(n_1) \\
+&= 40 \text{ seconds}
+\end{align*}$$
+
+So, when we double the size of our our matrices to
+$$2n \times 2n$$, the time becomes $$(2n)^3 = 8n^3$$.
+Thus, the runtime will be roughly 8 times as long.
+
+## Big-O Examples - Truncation Errors
+
+We can also use Big-O notation to describe the truncation error. A numerical method is called $$n$$-th order accurate if its
+truncation error $$E(h)$$ obeys $$E(h) = \mathcal{O}(h^n)$$.
+
+Consider solving an interpolation problem. We have an interval of
+length $$h$$ where our interpolant is valid and we know that our approximation
+is order $$\mathcal{O}(h^2)$$. What this means is that as we decrease h (the interval
+length), our error will decrease quadratically. Using the definition of Big-O,
+we know that $$\text{Error} = C \cdot h^2$$ where $$C$$ is some constant.
+
+In some cases, we may not know the exponent in $$E(h) = \mathcal{O}(h^n)$$.  We can estimate it using by computing the error at two different values of $$h$$.  Suppose we have two quantities, $$h_1 = 0.5$$ and $$h_2 = 0.25$$.  We compute the corresponding errors as $$E(h_1) = 0.125$$ and $$E(h_2) = 0.015625$$.  Then, since $$E(h) = \mathcal{O}(h^n)$$, we have:
+
+
+$$
+\begin{eqnarray}\frac{0.125}{0.015625}  &=\frac{E(h_1)}{E(h_2)} \\
+&\approx\frac{Ch_1^n}{Ch_2^n}\\
+&=\left(\frac{h_1}{h_2}\right)^n\\
+\end{eqnarray}
+$$
+
+$$
+\begin{eqnarray}
+\implies\log\left(\frac{0.125}{0.015625}\right)=
+n\log\left(\frac{h_1}{h_2}\right) =n\log\left(\frac{0.5}{0.25}\right)
+\end{eqnarray}
+$$
+
+Solving this equation for $$n$$, we obtain $$n = 3$$.
+

```

Comment:
do we need a specific function-example for errors? i don't think we do but the slides go pretty far into depth on taylor approximations for sin(x)